File: src/tests/file-rust/src/unsafe/fs/chdir.rs
// Copyright(c) The Maintainers of Nanvix.
// Licensed under the MIT License.

//==================================================================================================
// Imports
//==================================================================================================

use ::alloc::string::{
    String,
    ToString,
};
use ::syscall::unistd;

//==================================================================================================
// Standalone Functions
//==================================================================================================

/// Tests whether we can change the current working directory using `chdir()`.
pub fn test() {
    let target_directory: String = "src".to_string();

    // Save the current working directory.
    let saved_current_directory: String = match unistd::getcwd() {
        Ok(path) => path,
        Err(error) => panic!("{error:?}"),
    };

    // Change the current working directory to the target directory.
    if let Err(error) = unistd::chdir(&target_directory) {
        panic!("{error:?}");
    }

    // Ensure that the current working directory has changed.
    match unistd::getcwd() {
        Ok(current_directory) => {
            if current_directory == saved_current_directory {
                panic!("current working directory did not change");
            }
        },
        Err(error) => panic!("{error:?}"),
    };

    // Restore the saved current working directory.
    if let Err(error) = unistd::chdir(&saved_current_directory) {
        panic!("{error:?}");
    }

    // Ensure that the current working directory is restored.
    match unistd::getcwd() {
        Ok(current_directory) => {
            if current_directory != saved_current_directory {
                panic!("current working directory is not restored");
            }
        },
        Err(error) => panic!("{error:?}"),
    };
}

File: src/tests/file-c/chdir.c
/*
 * Copyright(c) The Maintainers of Nanvix.
 * Licensed under the MIT License.
 */

//==================================================================================================
// Imports
//==================================================================================================

#include <assert.h>
#include <fcntl.h>
#include <limits.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

//==================================================================================================
// Standalone Functions

// Tests whether we can change the current working directory.
void test_chdir(void)
{
    fprintf(stderr, "testing chdir() ... ");

    const char *dirname = "src";
    assert(strlen(dirname) <= NAME_MAX);

    char original_cwd[PATH_MAX];
    char new_cwd[PATH_MAX];

    // Get the current working directory.
    assert(getcwd(original_cwd, sizeof(original_cwd)) != NULL);

    // Change to the target directory.
    assert(chdir(dirname) == 0);

    // Verify the current working directory has changed.
    assert(getcwd(new_cwd, sizeof(new_cwd)) != NULL);
    assert(strcmp(new_cwd, original_cwd) != 0);

    // Restore the original working directory.
    assert(chdir(original_cwd) == 0);

    // Verify the current working directory is restored.
    assert(getcwd(new_cwd, sizeof(new_cwd)) != NULL);
    assert(strcmp(new_cwd, original_cwd) == 0);

    fprintf(stderr, "passed\n");
}
